src/storage/mvcc/reader/util.rs

```diff
@@ -0,0 +1,209 @@
+/// the safe `ts` will be returned.
+pub fn check_lock(key: &Key, ts: u64, lock: &Lock) -> Result<u64> {
+    if lock.ts > ts || lock.lock_type == LockType::Lock {
+        // Ignore lock when lock.ts > ts or lock's type is Lock
+        return Ok(ts);
+    }
+
+    let raw_key = key.raw()?;
+
+    if ts == ::std::u64::MAX && raw_key == lock.primary {
+        // When `ts == u64::MAX` (which means to get latest committed version for
+        // primary key), and current key is the primary key, we return the latest
+        // committed version.
+        return Ok(lock.ts - 1);
+    }
+
+    // There is a pending lock. Client should wait or clean it.
+    Err(Error::KeyIsLocked {
+        key: raw_key,
+        primary: lock.primary.clone(),
+        ts: lock.ts,
+        ttl: lock.ttl,
+    })
+}
+
+/// Reads user key's value in default CF according to the given write CF value
+/// (`write`).
+///
+/// Internally, there will be a `near_seek` operation with `allow_reseek == true`.
+///
+/// Notice that the value may be already carried in the `write` (short value). In this
+/// case, you should not call this function.
+///
+/// # Panics
+///
+/// Panics if there is a short value carried in the given `write`.
+///
+/// Panics if key in default CF does not exist. This means there is a data corruption.
+pub fn near_load_data_by_write<I>(
+    default_cursor: &mut Cursor<I>, // TODO: make it `ForwardCursor`.
+    user_key: &Key,
+    write: Write,
+    statistics: &mut Statistics,
+) -> Result<Value>
+where
+    I: Iterator,
+{
+    assert!(write.short_value.is_none());
+    let seek_key = user_key.clone().append_ts(write.start_ts);
+    default_cursor.near_seek(&seek_key, true, &mut statistics.data)?;
+    assert!(default_cursor.valid());
+    assert!(default_cursor.key(&mut statistics.data) == seek_key.encoded().as_slice());
+    statistics.data.processed += 1;
+    Ok(default_cursor.value(&mut statistics.data).to_vec())
+}
+
+/// Similar to `near_load_data_by_write`, but accepts a `BackwardCursor` and use
+/// `near_seek_for_prev` internally.
+pub fn near_reverse_load_data_by_write<I>(
+    default_cursor: &mut Cursor<I>, // TODO: make it `BackwardCursor`.
+    user_key: &Key,
+    write: Write,
+    statistics: &mut Statistics,
+) -> Result<Value>
+where
+    I: Iterator,
+{
+    assert!(write.short_value.is_none());
+    let seek_key = user_key.clone().append_ts(write.start_ts);
+    default_cursor.near_seek_for_prev(&seek_key, true, &mut statistics.data)?;
+    assert!(default_cursor.valid());
+    assert!(default_cursor.key(&mut statistics.data) == seek_key.encoded().as_slice());
+    statistics.data.processed += 1;
+    Ok(default_cursor.value(&mut statistics.data).to_vec())
+}
+
+/// Get Mvcc properties stored in each sstable's metadata.
+///
+// TODO: Move it out of mvcc::reader.
+pub fn get_mvcc_properties<S>(snapshot: &S, safe_point: u64) -> Option<MvccProperties>
```

Comment:
I think you should rename this function to `get_mvcc_properties_for_gc`.

Reply:
We can move it when we extracting gc.